### ISSUE TITLE
Rust exploration

### ISSUE THREAD

**Issue by higepon** [higepon/mosh#236]
Based on [Introduction - Writing Interpreters in Rust: a Guide](https://rust-hosted-langs.github.io/book/introduction.html) we explore if it's fun to rewrite Mosh in Rust.

## Goals
- We'll see if it's practically possible to rewrite Mosh in Rust with the current design (Compiler written in Scheme).
- We'll see if all basic building blocks work.
  - Scheme object with tag bits.
  - GC
  - VM
  - UTF-8
  - File I/O

## Non Goals
- Fully rewrite Mosh.
- Well designed Rust code.

## Milestones
- M1: Just build the example
- M2: Define Fixnum and it's predicate
- M3: Create simple VM to just evaluate constant
- M4: Update VM to run addtion
- M5: Come up with a rough idea of how we implement GC.

**Comment by higepon** [comment]
## M2 TODOs
- define and print `usize` number.
- print bitwise `and` for the number
- define unsafe raw_pointer
- add method to extract number value from the pointer
- add predicate method for the raw_pointer
- Revisit https://rust-hosted-langs.github.io/book/chapter-interp-tagged-ptrs.html
- Define next steps

**Comment by higepon** [comment]
### M3 and M4
- [ ] https://github.com/higepon/mosh/blob/dc716dc31b3e406f49161ea3cdfc9a8b4021e44d/rmosh/src/main.rs#L20

**Comment by higepon** [comment]
Need to read this.

**Comment by higepon** [comment]
## M5

**Comment by higepon** [comment]
This is good read but we should not assume this implementation is production quality.
https://github.com/ceronman/loxido

**Comment by higepon** [comment]
Looked into this https://higepon.hatenablog.com/entry/2022/12/03/160801.

**Comment by higepon** [comment]
Now I understand how enum in Rust works. I think it maybe okay implementing object system w/o tagged pointers.
Some of my tweets about it.
https://twitter.com/HigeponJa/status/1600265255317573637

**Comment by higepon** [comment]
### M7: identify very small Scheme program, compile it in Mosh and execute it in the Rust VM
- [ ] Just evaluate define and reference.
- [ ] 1 let.

**Comment by higepon** [comment]
https://github.com/higepon/mosh/blob/b07ad0a4c8add619bab7c58dae303369f69fbfca/rmosh/src/main.rs

**Comment by higepon** [comment]
## M9: Understand how GC works and trigger GC properly. Figure out how to test it.
- [ ] Mimic loxido print debug of gc
- [ ] figure out how to turn on/off the debug print

**Comment by higepon** [comment]
- [ ] Finish implementing all core instructions.
  - [ ] Stop using hand written instructions
    - [ ] Change vm.scm to compile-file and output it to stdout w/o optimization.
    - [ ] Write a scheme program to rewrite it Op::Style.
    - [ ] Test it with the existing vm tests.
  - [ ] Run compiled all-tests.scm
    - [ ] Convert 1 test to Op::
    - [ ] Pass the test with the following.
      - [ ] Make the free var stub with a dummy lambda + display.
      - [ ] Enable args and argc for functions implemented in Rust
      - [ ] Add instructions needed.
    - Repeat this. Until we can run most of them.
    - Update this milestones as we go.  


- [ ] Make the free vars same as Mosh
- [ ] Run compiled all-tests.scm
- [ ] Run the compiler in the VM

**Comment by higepon** [comment]
## Some notes for next steps
- Running compiled tests were very efficient to implement VM instructions and procs.
- The next big milestones are
  - Being able to run the compiler written in Scheme.
  - Parser (This should be done after the compiler).
- To be able to run the compiler we need
  - Being able to run long program
  - Compiled instructions of the compiler.
  - Base library written in Scheme.
  - Base library written in Rust.

**Comment by higepon** [comment]
## M11 Run as many tests in test-data.scm as possible
- [ ] Implements important primitives such as bytevector, vector.
- [ ] Skip test only if it's too early to implement and add todo comment there.

**Comment by higepon** [comment]
##M12 Being able to load small compiled program as baselib

**Comment by higepon** [comment]
### M13: Being able to run `(compile 3)`
- A couple of things I have to think through.
  - The compiler running on the VM is a list of Op.
  - The compiler doesn't return the list of Op but it does return
    - A list of flat insn vector => this should be converted into Op.
    - We also covert the insn num to Op.

**Comment by higepon** [comment]
## M14: Flatten a list of instructions
### Background
I think I made a wrong design choice on how we treat a list of instructions.
In C++ Mosh instruction is an object and a list of instruction is a list of object.
But in Rust Mosh we treat instruction as enum *with* value. Such as Constant(3) or Call(2). So they are not Object any more. And they can't be in a list of Object.
We found two major downsides.
- (A) Instruction and object are different Rust type and that complicates the code a lot. 
- (B) We have to convert a list of instruction made by compiler into Enum and adjust offset.

### Changes we'll make
- Compiler instruction is an Object. Specifically Object::Instruction(Op) where Op is Enum w/o operand.
- Remove all un-flatten code.

**Comment by higepon** [comment]
### M15: Simple Reader
- [ ] Investigate parser & helpers.
- [ ] Decide what to you.
- [ ] Read number
- [ ] Read symbol
- [ ] Read string
- [ ] Read dot pair
- [ ] Read pair
- [ ] Read vector
- [ ] And more :)

**Comment by higepon** [comment]
## GC todo
- Invoking GC in vm.alloc is not good design. It can cause memory error where the allocated object itself is freed because it's not rooted. We should think where to trigger gc.

**Comment by higepon** [comment]
Now rmosh can read and run a program.
```sh
$ cat hoge.scm 
(display ((lambda (a) (+ a 1)) 2))
$ ./target/debug/rmosh hoge.scm 
3
```

**Comment by higepon** [comment]
### M16 Make the reader better
- Run the all-tests.scm and improve the reader.

**Comment by higepon** [comment]
Now rmosh can run simple R6RS program with some errors :)
```
(import (rnrs))

(display "Hello")
```

**Comment by higepon** [comment]
## M16
- Load `(import (mosh))`.
- Load `(import (mosh file)`.
- Enable `fasl`

**Comment by higepon** [comment]
### Weird bug on loading psyntax
```
free_var=source-info called
free_var=procedure? called
thread 'main' panicked at 'Not a Object::Closure but #<vox #<closure 0xaaaaf6956ca0>>', src/objects.rs:238:13
```

It is happening in parse-library
```
  (define parse-library
    (lambda (e)
      (format (current-error-port) "parse-library0" )  <==
      (syntax-match e ()
```

In `refer free push`, dc is supposed to be closure but was closure.

### Next steps
- [ ] commit before the big change
- [ ] Implement writer which can print cyclic reference object.
- [ ] Print stack and vm instructions for the bug and check what's wrong.

**Comment by higepon** [comment]
Now rmosh can run the following program.
```scheme
(import (scheme base))
(import (scheme write))
(import (only (srfi :1) list-ref))
(import (mosh control))
(display 3)
(newline)
```

**Comment by higepon** [comment]
Now rmosh can automatically serialize library as mosh does.
```
serialize-library /root/mosh.git/lib/mosh/control.ss
...serialize-library /root/mosh.git/lib/scheme/write.mosh.sls
...serialize-library /root/mosh.git/lib/r7b-impl/write.mosh.sls
...serialize-library /root/mosh.git/lib/scheme/base.mosh.sls
...serialize-library /root/mosh.git/lib/r7b-impl/base.sls
...serialize-library /root/mosh.git/lib/r7b-util/case.sls
...serialize-library /root/mosh.git/lib/r7b-util/syntax-rules.sls
...serialize-library /root/mosh.git/lib/r7b-util/char-ready.sls
...serialize-library /root/mosh.git/lib/r7b-impl/division.sls
...serialize-library /root/mosh.git/lib/srfi/%3a43.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a43/vectors.sls
...serialize-library /root/mosh.git/lib/srfi/%3a13.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a13/strings.sls
...serialize-library /root/mosh.git/lib/srfi/%3a14/char-sets.sls
...serialize-library /root/mosh.git/lib/srfi/private/include.sls
...serialize-library /root/mosh.git/lib/srfi/private/include/compat.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/private/let-opt.sls
...serialize-library /root/mosh.git/lib/srfi/%3a8/receive.sls
...serialize-library /root/mosh.git/lib/srfi/%3a1.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a1/lists.sls
...serialize-library /root/mosh.git/lib/srfi/i39.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a39.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/i9.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a9.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a9/records.sls
...serialize-library /root/mosh.git/lib/srfi/i23.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a23.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a23/error.sls
...serialize-library /root/mosh.git/lib/srfi/%3a39/parameters.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/i6.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a6.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a6/basic-string-ports.sls
...serialize-library /root/mosh.git/lib/srfi/%3a6/basic-string-ports/compat.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/i0.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a0.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/%3a0/cond-expand.sls
...serialize-library /root/mosh.git/lib/srfi/private/registry.sls
...serialize-library /root/mosh.git/lib/srfi/private/platform-features.mosh.sls
...serialize-library /root/mosh.git/lib/srfi/private/OS-id-features.sls
```

**Comment by higepon** [comment]
### M17
Start working on R7RS.
`cargo run ../tests/r7rs/import-all.scm `

**Comment by higepon** [comment]
call/cc worked.
```
(import (rnrs))

(display (call/cc (lambda (c)  (c 3))))
(newline)
```

**Comment by higepon** [comment]
Anyone interested in writing R7RS procedures in Rust?
I'm now trying to run all tests in ../tests/r7rs/r7rs-tests.scm.

We have to implement ~400 procedures in https://github.com/higepon/mosh/blob/bigint/rmosh/src/procs.rs.
For example I recently implemented "+" as follows.
```
fn number_add(vm: &mut Vm, args: &mut [Object]) -> Object {
    let name: &str = "+";
    let argc = args.len();
    if argc == 0 {
        Object::Fixnum(0)
    } else if argc == 1 {
        if args[0].is_number() {
            args[0]
        } else {
            panic!("{}: number required but got {}", name, args[0])
        }
    } else {
        let mut ret = Object::Fixnum(0);
        for arg in args.iter() {
            ret = numbers::add(&mut vm.gc, ret, *arg);
        }
        ret
    }
}
```

**Comment by okuoku** [comment]
I wish i had enough Rust-fu to help you...

(But anyway I'm really positive to switch Mosh implementation to Rust; I need a few months familiarise myself for it though)

**Comment by higepon** [comment]
Haha thanks. If you take a closer look. rmosh is just a copy of Mosh. 
- VM instructions
- Compiler
- object system
are the same.

Anyway ping me when you have some time.
So far I implemented 200 procedures and need 400 more :)

**Comment by higepon** [comment]
## Exception design
Exceptions can be raised in the following cases. 
- (A) When executing VM instruction. eg) div by zero in NumberDiv instruction.
- (B) When running one of the scheme procedures written in Rust. eg) Try to access car of nil.
- (C) When raise is called.
Here we discuss how we implement Scheme exceptions in Rust.

### How we do it in C++ Mosh?
- (A) `callAssertionViolationAfter` eventually calls `theVM->throwException(condition)` then calls `longjmp`.
  - Basically before calling VM loop we `setjmp` and then come back there using `longjmp`.

**Comment by higepon** [comment]
- Test rmosh using miri regularly.

**Comment by higepon** [comment]
`[  FAILED  ] 1136 passed, 70 failed, 10 skipped.`
70 failures more!

**Comment by higepon** [comment]
```sh
$ time cargo run  ../tests/r7rs/r7rs-tests.scm
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/rmosh ../tests/r7rs/r7rs-tests.scm`
WARNING open-file-output-port: ["/root/.rmosh/prefix.txt", #< simple-struct #< simple-struct type:record-type-descriptor enum-set, #f, enum-set-uid, #f, #f, ((#t . type) (#t . members))> #< simple-struct #< simple-struct type:record-type-descriptor enum-type, #f, enum-type-uid, #f, #f, ((#t . universe) (#t . indexer))> (no-create no-fail no-truncate), #<closure (lambda symbol) 0xaaaab96a4300>>, (no-fail)>, block, "TODO: native-transcoder"] silently ignored
eqv_rat 0.1 0.3
  (SKIPPED (let () (define-syntax elli-lit-1 (syntax-rules ... (...) ((_ x) '(x ...)))) (test '(100 ...) (elli-lit-1 100))))
  (SKIPPED (test #f (string->number "1 2")))
  (SKIPPED (test #t (read-error? (guard (exn (else exn)) (read (open-input-string "\""))))))
  (SKIPPED (test 'error (let ((in (open-input-string "abc"))) (close-input-port in) (guard (exn (else 'error)) (read-char in)))))
  (SKIPPED (test #t (char-ready? (open-input-string "42"))))
  (SKIPPED (test "((1 2 3) (1 2 3))" (let ((out (open-output-string)) (x (list 1 2 3))) (write (list x x) out) (get-output-string out))))
  (SKIPPED (test "line 1\n\nline 3\n" (read (open-input-string "\"line 1\\ \t \n \t \n\nline 3\n\""))))
  (SKIPPED (test-precision "-1.7976931348623157e+308" "-inf.0"))
  (SKIPPED (test-precision "1.976262583364986e-323" "1.97626258336499e-323" "2.0e-323" "0.0"))
  (SKIPPED (test-precision "1.7976931348623157e+308" "+inf.0"))

[  PASSED  ] 1210 passed, 10 skipped. 
```

**Comment by higepon** [comment]
## M18: Pass R6RS tests
- [ ] Clean up stack trace
- [ ] Show file name
- [ ] Show correct line number

**Comment by higepon** [comment]
- [ ] Support Ratnum with Bigint
- [ ] Support Bignum read in number_reader.lalrpop

**Comment by higepon** [comment]
Passed
```sh
target/debug/rmosh ../tests/r6rs-test-suite/tests/r6rs/run/base.sps --loadpath=../tests/r6rs-test-suite/`
```

**Comment by higepon** [comment]
Passed all r6rs tests :)

**Comment by higepon** [comment]
### todo for rmosh 0.0.1
- install library files as part of `cargo install`.
- set up CI

**Comment by higepon** [comment]
Released at https://crates.io/crates/rmosh/.

If you have Rust environment already. `cargo install rmosh` is all you have to do.

**Comment by higepon** [comment]
## Remove ReadError and use SchemeError
We added ReadError struct to handle read related errors. But it turned out it was not necessary and adding one more unnecessary layer of error conversion.

## Steps
- Add SchemeError enum
- Change lexer to return SchemeError.
- Change reader to return SchemeError.
- Change `read` to return SchemeError
- Propergate the error as correspodning Scheme error.
- Test if it works.
- Do the same for number_iter, number_reader.

**Comment by higepon** [comment]
https://github.com/higepon/mosh/pull/255

**Comment by higepon** [comment]
https://crates.io/crates/rmosh/ 0.0.11 is out.
Passing more tests.

## TODO
- `(mosh socket)`
- regexp

regexp should be very straightforward to implement, but not sure about socket.
Please let me know if you have any interests.

**Comment by higepon** [comment]
Forgot to mention that rmosh is one binary std libraries embedded in binary.

**Comment by higepon** [comment]
Now rmosh passes regexp tests and more https://github.com/higepon/mosh/blob/master/rmosh/Makefile.
Only remaining work is probably socket.

**Comment by higepon** [comment]
Simple client socket is now supported. http-get worked. I don't do further socket work until it becomes really necessary.
Will push the release at https://crates.io/crates/rmosh/.

I think it's kinda code completed. I'd love any feedback from Scheme community.
Is there any places I can get feedback other than https://www.reddit.com/r/scheme/comments/1296vjc/introducing_rmosh_a_rust_implementation_of_mosh/?
